packages/beacon-node/src/chain/rewards/blockRewards.ts

```diff
@@ -0,0 +1,140 @@
+
+  return {
+    proposerIndex: block.proposerIndex,
+    total,
+    attestations: blockAttestationReward,
+    syncAggregate: syncAggregateReward,
+    proposerSlashings: blockProposerSlashingReward,
+    attesterSlashings: blockAttesterSlashingReward,
+  };
+}
+
+/**
+ * TODO: Calculate rewards received by block proposer for incuding attestations.
+ */
+function computeBlockAttestationRewardPhase0(
+  _block: phase0.BeaconBlock,
+  _state: CachedBeaconStatePhase0
+): SubRewardValue {
+  throw new Error("Unsupported fork! Block attestation reward calculation is not yet available in phase0");
+}
+
+/**
+ * Calculate rewards received by block proposer for incuding attestations since Altair.
+ * Reuses `processAttestationsAltair()`. Has dependency on RewardCache
+ */
+function computeBlockAttestationRewardAltair(
+  block: altair.BeaconBlock,
+  state: CachedBeaconStateAltair
+): SubRewardValue {
+  const fork = state.config.getForkSeq(block.slot);
+  const {attestations} = block.body;
+
+  processAttestationsAltair(fork, state, attestations, false);
+
+  return state.proposerRewards.attestations;
+}
+
+function computeSyncAggregateReward(block: altair.BeaconBlock, state: CachedBeaconStateAltair): SubRewardValue {
+  if (block.body.syncAggregate !== undefined) {
+    const {syncCommitteeBits} = block.body.syncAggregate;
+    const {syncProposerReward} = state.epochCtx;
+
+    return syncCommitteeBits.getTrueBitIndexes().length * Math.floor(syncProposerReward); // syncProposerReward should already be integer
+  } else {
+    return 0; // phase0 block does not have syncAggregate
+  }
+}
+/**
+ * Calculate rewards received by block proposer for include proposer slashings.
+ * All proposer slashing rewards go to block proposer and none to whistleblower as of Deneb
+ */
+function computeBlockProposerSlashingReward(
+  block: allForks.BeaconBlock,
+  state: CachedBeaconStateAllForks
+): SubRewardValue {
+  let proposerSlashingReward = 0;
+
+  for (const proposerSlashing of block.body.proposerSlashings) {
+    const offendingProposerIndex = proposerSlashing.signedHeader1.message.proposerIndex;
+    const offendingProposerBalance = state.validators.get(offendingProposerIndex).effectiveBalance;
+
+    proposerSlashingReward += Math.floor(offendingProposerBalance / WHISTLEBLOWER_REWARD_QUOTIENT);
+  }
+
+  return proposerSlashingReward;
+}
+
+/**
+ * Calculate rewards received by block proposer for include attester slashings.
+ * All attester slashing rewards go to block proposer and none to whistleblower as of Deneb
+ */
+function computeBlockAttesterSlashingReward(
+  block: allForks.BeaconBlock,
+  state: CachedBeaconStateAllForks
+): SubRewardValue {
+  let attesterSlashingReward = 0;
+
+  for (const attesterSlashing of block.body.attesterSlashings) {
+    for (const offendingAttesterIndex of getAttesterSlashableIndices(attesterSlashing)) {
+      const offendingAttesterBalance = state.validators.get(offendingAttesterIndex).effectiveBalance;
```

Comment:
same to the above, need to get `validators` once, and use `getReadOnly()` api, see sample usages in `state-transition` package

Reply:
Thanks!